rclcpp/include/rclcpp/subscription_options.hpp

```diff
@@ -0,0 +1,167 @@
+
+#include "rclcpp/visibility_control.hpp"
+
+namespace rclcpp
+{
+
+struct SubscriptionEventCallbacks
+{
+  // NOTE(emersonknapp): empty placeholder for future additions of QoS Event callbacks
+};
+
+template<typename Alloc = std::allocator<void>>
+class SubscriptionOptions
+{
+public:
+  /// Create SubscriptionOptions with default values, optionally specifying the allocator to use.
+  /**
+   * Default values for the node options:
+   *
+   *   - deadline_callback = nullptr
+   *   - liveliness_callback = nullptr
+   *   - lifespan_callback = nullptr
+   *   - subscription_qos_profile = rmw_qos_profile_default
+   *   - ignore_local_publications = false
+   *   - allocator = nullptr
+   *
+   * \param[in] allocator allocator to use in construction of SubscriptionOptions.
+   */
+  RCLCPP_PUBLIC
+  SubscriptionOptions() = default;
+
+  RCLCPP_PUBLIC
+  explicit SubscriptionOptions(const rmw_qos_profile_t & qos_profile)
+  : qos_profile_(qos_profile) {}
+
+  /// Destructor.
+  RCLCPP_PUBLIC
+  virtual
+  ~SubscriptionOptions() = default;
+
+  /// Copy constructor.
+  RCLCPP_PUBLIC
+  SubscriptionOptions(const SubscriptionOptions & other) = default;
+
+  /// Assignment operator.
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
+  operator=(const SubscriptionOptions & other) = default;
+
+
+  /// Return a reference to the subscription_qos_profile QoS.
+  RCLCPP_PUBLIC
+  const rmw_qos_profile_t &
+  qos_profile() const
+  {
+    return qos_profile_;
+  }
+
+  RCLCPP_PUBLIC
+  rmw_qos_profile_t &
+  qos_profile()
+  {
+    return qos_profile_;
+  }
+
+  /// Set the subscription_qos_profile QoS, return this for parameter idiom.
+  /**
+   * The QoS settings to be used for the subscription
+   */
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
+  qos_profile(const rmw_qos_profile_t & subscription_qos_profile)
+  {
+    qos_profile_ = subscription_qos_profile;
+    return *this;
+  }
+
+  RCLCPP_PUBLIC
+  size_t &
+  qos_history_depth()
+  {
+    return qos_profile_.depth;
+  }
+
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
+  qos_history_depth(size_t depth)
+  {
+    qos_profile_.depth = depth;
+    return *this;
+  }
+
+  RCLCPP_PUBLIC
+  const SubscriptionEventCallbacks &
+  event_callbacks() const
+  {
+    return callbacks_;
+  }
+
+
+  RCLCPP_PUBLIC
+  bool
+  ignore_local_publications() const
+  {
+    return ignore_local_publications_;
+  }
+
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
+  ignore_local_publications(bool ignore)
+  {
+    ignore_local_publications_ = ignore;
+    return *this;
+  }
+
+  /// Return the std::shared_ptr<Alloc> to be used.
+  RCLCPP_PUBLIC
+  std::shared_ptr<Alloc>
+  allocator() const
+  {
+    return allocator_;
+  }
+
+  RCLCPP_PUBLIC
+  SubscriptionOptions &
+  allocator(std::shared_ptr<Alloc> allocator)
+  {
+    allocator_ = allocator;
+    return *this;
+  }
+
+private:
+  // IMPORTANT: if any of these default values are changed, please update the
+  // documentation in this class.
+
+  SubscriptionEventCallbacks callbacks_;
```

Comment:
Using a unique_ptr here would spare you the need for writing copy-constructor and assignment constructor.